ostd/src/arch/riscv/irq/chip/plic.rs

```diff
@@ -0,0 +1,358 @@
+                        .expect("Failed to find '/cpus' node")
+                        .children()
+                        .filter(|node| node.name.split('@').next().unwrap() == "cpu")
+                        .collect::<Vec<_>>();
+
+                    plic_node
+                        .property("interrupts-extended")
+                        .expect("Failed to read 'interrupts-extended' property from PLIC node")
+                        .value
+                        .chunks_exact(8)
+                        .enumerate()
+                        .filter_map(|(idx, chunk)| {
+                            let target = [
+                                u32::from_be_bytes(chunk[0..4].try_into().unwrap()),
+                                u32::from_be_bytes(chunk[4..8].try_into().unwrap()),
+                            ];
+
+                            if target[1] != 0x09 {
+                                return None;
+                            }
+
+                            let hart_id = cpu_nodes.iter().find_map(|node| {
+                                node.children()
+                                    .find(|child| {
+                                        child
+                                            .compatible()
+                                            .is_some_and(|c| c.all().any(|s| s == "riscv,cpu-intc"))
+                                            && child.property("phandle").is_some_and(|ph| {
+                                                ph.as_usize().unwrap() as u32 == target[0]
+                                            })
+                                    })
+                                    .and_then(|_| node.property("reg").and_then(|p| p.as_usize()))
+                            })?;
+
+                            Some((hart_id as u32, idx as u32))
+                        })
+                        .collect()
+                };
+
+                Self {
+                    phandle,
+                    io_mem: io_mem_builder.reserve(range),
+                    hart_to_target_mapping,
+                    interrupt_number_mappings: (0..num_interrupt_sources)
+                        .map(|_| None)
+                        .collect::<Vec<Option<u8>>>()
+                        .into_boxed_slice(),
+                }
+            })
+            .collect()
+    }
+
+    pub(super) fn map_interrupt_source_to(
+        &mut self,
+        interrupt_source: u32,
+        irq_line: &IrqLine,
+    ) -> Result<()> {
+        // An interrupt source number of 0 is reserved to mean “no interrupt”.
+        // Reference: <https://github.com/riscv/riscv-plic-spec>.
+        assert_ne!(interrupt_source, 0);
+
+        match self.interrupt_number_mappings[interrupt_source as usize] {
+            None => {
+                self.interrupt_number_mappings[interrupt_source as usize] = Some(irq_line.num());
+                Ok(())
+            }
+            Some(_) => Err(Error::InvalidArgs),
```

Comment:
https://github.com/asterinas/asterinas/blob/7e7ba43cdfc572cd2ddeecc86a5951bbd200d692/ostd/src/arch/x86/irq/chip/ioapic.rs#L90-L92

This is `Error::AccessDenied` in x86-64.